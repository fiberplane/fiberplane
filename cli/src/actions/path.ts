import type { Context } from "@/context";
import { text } from "@clack/prompts";

// TODO: validate path
export async function promptPath(ctx: Context) {
  try {
    const placeholder = "./spooking-honc";
    const result = await text({
      message: "Where should we create your project? (./relative-path)",
      placeholder,
      defaultValue: placeholder,
      validate: (value) => {
        if (value !== "" && value[0] !== ".") {
          return "Please enter a relative path.";
        }
        return undefined;
      },
    });

    if (typeof result === "string") {
      if (result === "") {
        ctx.path = placeholder;
      } else {
        ctx.path = result;
      }
    }

    return result;
  } catch (error) {
    return error;
  }
}
