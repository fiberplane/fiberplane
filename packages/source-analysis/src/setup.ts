import { RoutesMonitor } from "./RoutesMonitor";
import { type Logger, logger } from "./logger";

export function createRoutesMonitor(
  projectRoot: string,
  overrideLogger?: Logger,
): RoutesMonitor {
  if (overrideLogger) {
    logger.logger = overrideLogger;
  }

  return new RoutesMonitor(projectRoot);
}
