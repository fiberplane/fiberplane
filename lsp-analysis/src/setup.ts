import path from "node:path";
import { extractRouteTrees } from "./findHono";
import { getTsLib, startServer } from "./server";
import type { TsType } from "./types";
import { Watcher } from "./watcher";

export function setupMonitoring(projectRoot: string) {
  const ts: TsType = getTsLib(projectRoot);

  // Only watch the src directory (if it exists)
  const possibleLocation = path.resolve(path.join(projectRoot, "src"));
  const location = ts.sys.directoryExists(possibleLocation)
    ? possibleLocation
    : projectRoot;
  const watcher = new Watcher(location);

  const fileMap: Record<
    string,
    {
      version: number;
      content: string;
    }
  > = {};

  function getFileInfo(fileName: string) {
    return fileMap[fileName];
  }

  function getFileNames() {
    return Object.keys(fileMap);
  }

  const server = startServer({
    getFileInfo,
    getFileNames,
    location: projectRoot,
    ts,
  });

  watcher.on("fileAdded", (event) => {
    fileMap[event.payload.fileName] = {
      version: 0,
      content: event.payload.content,
    };
    server.getProgram()?.getSourceFile(event.payload.fileName);
  });

  watcher.on("fileUpdated", (event) => {
    fileMap[event.payload.fileName] = {
      version: fileMap[event.payload.fileName].version + 1,
      content: event.payload.changes[0].text,
    };
  });

  watcher.on("fileRemoved", (event) => {
    delete fileMap[event.payload.fileName];
  });

  return {
    watcher,
    findHonoRoutes: () => extractRouteTrees(server, ts),
    teardown: () => {
      watcher.teardown();
      server.dispose();
    },
  };
}
