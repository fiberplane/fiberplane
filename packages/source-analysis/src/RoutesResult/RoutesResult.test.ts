import path from "node:path";
import { assert, expect, test } from "vitest";
import { createRoutesMonitor } from "../setup";

test("hono-factory", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/hono-factory");
  const monitor = createRoutesMonitor(absolutePath);
  monitor.autoCreateResult = false;
  try {
    await monitor.start();
    const factory = monitor.updateRoutesResult();
    assert(factory.rootId);

    const request = new Request("http://localhost/", { method: "GET" });
    const response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(3);

    const history = factory.getHistory();
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();

    // Expect the app from the factory file to be visited
    expect(
      history.find((item) => item.startsWith("ROUTE_TREE:factory.ts@")),
    ).toBeTruthy();
    expect(
      history.find((item) => item.startsWith("ROUTE_ENTRY:factory.ts@")),
    ).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    monitor.stop();
  }
});

test("barrel-files", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/barrel-files");
  const monitor = createRoutesMonitor(absolutePath);
  monitor.autoCreateResult = false;
  try {
    await monitor.start();
    const factory = monitor.updateRoutesResult();
    assert(factory.rootId);

    let request = new Request("http://localhost/", { method: "GET" });
    let response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
    // Reset the history
    factory.resetHistory();

    request = new Request("http://localhost/user/1", { method: "GET" });
    response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    monitor.stop();
  }
});

test("import-as", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/import-as");
  const monitor = createRoutesMonitor(absolutePath);
  monitor.autoCreateResult = false;
  try {
    await monitor.start();
    const factory = monitor.updateRoutesResult();
    assert(factory.rootId);

    let request = new Request("http://localhost/user/1", { method: "GET" });
    let response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();

    factory.resetHistory();
    request = new Request("http://localhost/user/1/profile", { method: "GET" });
    response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    monitor.stop();
  }
});

test("multiple", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/multiple");
  const monitor = createRoutesMonitor(absolutePath);
  monitor.autoCreateResult = false;
  try {
    await monitor.start();
    const factory = monitor.updateRoutesResult();
    assert(factory.rootId);

    let request = new Request("http://localhost/", { method: "GET" });
    let response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();

    factory.resetHistory();
    // Try to visit a route that doesn't exist
    request = new Request("http://localhost/hello-world", { method: "GET" });
    response = await factory.currentApp.fetch(request);
    expect(response.status).toEqual(404);
    expect(factory.getHistoryLength()).toBe(1);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();

    factory.resetHistory();
    // Try to visit a route that doesn't exist
    request = new Request("http://localhost/hello-world", { method: "POST" });
    response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    monitor.stop();
  }
});

test("zod-openapi", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/zod-openapi");
  const monitor = createRoutesMonitor(absolutePath);
  monitor.autoCreateResult = false;
  try {
    await monitor.start();
    const factory = monitor.updateRoutesResult();
    assert(factory.rootId);

    let request = new Request("http://localhost/users", { method: "POST" });
    let response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();

    factory.resetHistory();
    // Try to visit another route
    request = new Request("http://localhost/users", { method: "GET" });
    response = await factory.currentApp.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(factory.rootId)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();

  } finally {
    monitor.stop();
  }
});
