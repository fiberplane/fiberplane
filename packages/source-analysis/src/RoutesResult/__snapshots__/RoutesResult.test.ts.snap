// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barrel-files 1`] = `
"/* index.ts */
const app = new Hono();

app.get("/", (c) => c.text("Hello, Hono!")
/* EOF: index.ts */"
`;

exports[`barrel-files 2`] = `
"/* index.ts */
import { cors } from "hono/cors";
import { getUser } from "./db";
const app = new Hono();

app.get("/user/1", cors()async (c) => {
  // await getUser();
  const user = await getUser();
  return c.json(user);
}
/* EOF: index.ts */
/* db/index.ts */

export { getUser } from "./getUser"
/* EOF: db/index.ts */
/* db/getUser.ts */
import { measure } from "@fiberplane/hono-otel";
import { sleep } from "../utils";
import { User } from "../types";

const DEFAULT_USER_NAME = "John Doe"

const DEFAULT_EMAIL = "john@doe.com"

export const getUser = measure("getUser", async () => {
  await sleep(100);
  const value: User = {
    name: DEFAULT_USER_NAME,
    email: DEFAULT_EMAIL,
  };
  return value;
});
/* EOF: db/getUser.ts */
/* utils.ts */

export async function sleep(duration: number) {
  return new Promise((resolve) => setTimeout(resolve, duration));
}
/* EOF: utils.ts */
/* types.ts */

export type Email = string;

export type User = {
  name: string;
  email: Email;
};
/* EOF: types.ts */"
`;

exports[`hono-factory 1`] = `
"/* index.ts */
const app = new Hono();
/* EOF: index.ts */
/* factory.ts */
const app = new Hono();

app.get("/", (c) => c.text("Hello, Hono!")
/* EOF: factory.ts */"
`;

exports[`import-as 1`] = `
"/* index.ts */
import { cors } from "hono/cors";
import * as db from "./db";
const app = new Hono();

app.get("/user/1", cors()async (c) => {
  const user = await db.getUser();
  return c.json(user);
}
/* EOF: index.ts */
/* db.ts */
import { measure } from "@fiberplane/hono-otel";
import { path } from "node:path";

const sleep = (duration = 100) =>
  new Promise((resolve) => setTimeout(resolve, duration))

export const getUser = measure("getUser", async () => {
  await sleep();

  // Do something silly with the path module
  const parent = path.resolve(__dirname, "..");
  console.log("parent folder", parent);
  const value = {
    name: "John Doe",
    email: "john@doe.com",
  };
  return value;
});
/* EOF: db.ts */"
`;

exports[`import-as 2`] = `
"/* index.ts */
import { cors } from "hono/cors";
import { getProfile as getUserProfile } from "./db";
const app = new Hono();

app.get("/user/1/profile", cors()async (c) => {
  const profile = await getUserProfile();
  return c.json(profile);
}
/* EOF: index.ts */
/* db.ts */
import { measure } from "@fiberplane/hono-otel";
import { path } from "node:path";

const sleep = (duration = 100) =>
  new Promise((resolve) => setTimeout(resolve, duration))

export const getUser = measure("getUser", async () => {
  await sleep();

  // Do something silly with the path module
  const parent = path.resolve(__dirname, "..");
  console.log("parent folder", parent);
  const value = {
    name: "John Doe",
    email: "john@doe.com",
  };
  return value;
});

export async function getProfile() {
  const user = await getUser();
  await sleep(10);
  return {
    ...user,
    image: "https://xsgames.co/randomusers/avatar.php?g=pixel",
  };
}
/* EOF: db.ts */"
`;

exports[`multiple 1`] = `
"/* other.ts */
const app = new Hono();

app.get("/", (c) => {
  return c.json({});
}
/* EOF: other.ts */"
`;

exports[`multiple 2`] = `
"/* other.ts */
const app = new Hono();
/* EOF: other.ts */"
`;

exports[`multiple 3`] = `
"/* other.ts */
const app = new Hono();

app.post("/hello-world", (c) => c.json({ hello: "world" })
/* EOF: other.ts */"
`;
